src/Domain/DonationService.php

```diff
@@ -455,12 +458,41 @@ private function updateDonationFeesFromConfirmationToken(
         $token = $this->stripe->retrieveConfirmationToken($tokenId);
 
         /** @var StripeObject&object{
-         *     card: null|object{country: string, brand: string},
+         *     card: null|object{country: string, brand: string, fingerprint: string},
          *     pay_by_bank: null|StripeObject
          * } $paymentMethodPreview
          */
         $paymentMethodPreview = $token->payment_method_preview;
 
+        $card = $paymentMethodPreview->card;
+
+        if ($card) {
+            if ($card->fingerprint && $this->redis->exists('card-fingerprint-' . $card->fingerprint) === 0) {
+                $stripeCustomerId = $donation->getPspCustomerId()?->stripeCustomerId;
+                \assert(\is_string($stripeCustomerId));
+
+                $rateLimitId = 'confirm-donation-' . $stripeCustomerId;
+                $rateLImiter = (new RateLimiterFactory([
+                    'id' => $rateLimitId,
+                    'policy' => 'token_bucket',
+                    'limit' => 5,
+                    'rate' => ['interval' => '30 minutes'],
+                ], $this->rateLimiterstorage))->create();
```

Comment:
I'm not sure I see why we need to work with the underlying Redis directly.

Could we pass `key` to `create()` here and indeed inject a (differently aliased) `RateLimiterFactory`, so that the unchanging config lives in dependencies and everything works more like `DonationService::doCreateDonation()`?

Reply:
ah yes we could - I didn't see that `create` takes an argument. Yes I agree in that case it's better to put the rate limiter in the dependencies and inject it. Will do.

Also hadn't realised on Wednesday that we are already using the Symfony Rate Limiter.

But do you think using redis directly to check whether the card is a new one or not makes sense, since we only want to limit use of new cards, not retries of the same card they've tried before?

Reply:
Ah yes, I think I was temporarily confused about the card fingerprints when I first reviewed. That probably does make sense since it's sort of the opposite of how rate limiting strategies normally work (more tries are allowed if in the _same_ value bucket) so is unlikely to be covered out the box. And then the rate limiting bucket itself is based on customer ID.